addons/point_of_sale/models/pos_session.py

```diff
@@ -313,6 +302,522 @@ def action_pos_session_close(self):
+        rounding_error_line_vals = self._get_rounding_error_line_vals(account_move, session, line_args, is_using_company_currency)
+        return self.env['account.move.line'].create(line_args + rounding_error_line_vals)
+
+    @api.model
+    def _validate_cash_statement(self, session, statement, statement_payments):
+        """ Validates each statement of the given session then returns
+            cash account.bank.statement.line records.
+        """
+        if not statement_payments:
+            return self.env['account.bank.statement.line']
+        if abs(session.cash_register_difference) > statement.journal_id.amount_authorized_diff:
+            # The pos manager can close statements with maximums.
+            if not self.user_has_groups("point_of_sale.group_pos_manager"):
+                raise UserError(_(
+                    "Your ending balance is too different from the theoretical cash closing (%.2f),"
+                    "the maximum allowed is: %.2f. You can contact your manager to force it."
+                ) % (session.cash_register_difference, statement.journal_id.amount_authorized_diff))
+        cash_statement_lines = self._create_cash_statement_lines(session, statement, statement_payments)
+        statement_line_ids = cash_statement_lines.ids
+        if not session.config_id.cash_control:
+            statement.write({'balance_end_real': statement.balance_end})
+        statement.button_confirm_bank()
+        return self.env['account.bank.statement.line'].browse(statement_line_ids)
+
+    @api.model
+    def _create_cash_statement_lines(self, session, statement, statement_payments):
+        session_currency = session.currency_id
+        company = session.company_id
+        def create_vals(payment_method):
+            # create statement line for each payment method
+            payments = statement_payments.filtered(lambda p: p.payment_method_id == payment_method)
+            currency = payment_method.cash_journal_id.currency_id or company.currency_id
+            amount_converter = lambda payment: session_currency._convert(payment.amount, currency, company, payment.payment_date)
+            return {
+                'date': fields.Date.context_today(self),
+                'amount': sum(payments.mapped(amount_converter)) if payments else 0.0,
+                'name': session.name,
+                'statement_id': statement.id,
+                'account_id': payment_method.receivable_account_id.id,
+            }
+
+        cash_payment_methods = [pm for pm in session.payment_method_ids if pm.is_cash_count and (pm.cash_journal_id == statement.journal_id)]
+        statement_lines_args = [create_vals(pm) for pm in cash_payment_methods]
+        return self.env['account.bank.statement.line'].create([args for args in statement_lines_args if args.get('amount')])
+
+    @api.model
+    def _get_invoiced_receivable_lines_vals(self, account_move, invoiced_orders, is_using_company_currency):
```

Comment:
I think we should simplify this methods, and others a lot,
making it @api.multi, self.ensure_one(), and get the invoiced orders from a filtered in this method.
We should do this kind of code for all other move_line creation.

invoiced_orders = self.order_ids.filtered(lambda o: o.account_move)
keys = ['partner_id.property_account_receivable_id.id']
grouped_invoiced_orders = groupby(sort(invoiced_orders, key=itemgetter(*keys)), key=itemgetter(*keys))
for account_id, orders in grouped_invoiced_orders:
    total_amount = sum(order.amount_total for order in orders)
    #change total amount if currency different
    if float_compare(total_amount, 0) > 0:
        debit = 0
        credit = amount
    else:
        debit = amount
        credit = 0
    lines.append({
        'account_id': account_id,
        'move_id': account_move.id,
        'name': 'From invoiced orders',
        'debit': debit,
        'credit': credit,
    })
return lines

Reply:
I totally agree with this comment that this function and the others related to it should be simplified.